Check for username, ip, or registry on faas push

BurtonR opened the pull request:
Signed-off-by: Burton Rheutan <rheutan7@gmail.com>

<!--- Provide a general summary of your changes in the Title above -->
Added a check before calling `docker push` to ensure that the `image` field in the `.yml` file contains a repository by looking for `/`.

## Description
<!--- Describe your changes in detail -->
Added new function to `push.go` to validate that the image contains a repository by looking for `/` in the image string

Added output message with list of failed function names with message to indicate the failure (and an example of the correct usage)

I also moved the `"Pushing function1 done"` message into the else block that calls the `docker push` command. I thought it confusing to have that message show when the `push` never actually happened...

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
- [X] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))

Closes #329 

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Built the CLI locally and ran `./faas-cli push -f test.yml` with 3 functions defined. One with a repository `burtonr/movies:0.2`, one with `skip_build: true`, and one without a repository `ecs-cli:latest`. Verified that the push was blocked (in both synchronous and `--parallel`). Updated the `.yml` file to ensure both functions without `skip_build` had the repository in the image. Ran `./faas-cli push -f test.yml` again, and ensured that the push happened. See example `.yml` below. 

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [X] I have signed-off my commits with `git commit -s`
- [ ] I have added tests to cover my changes.
- [X] All new and existing tests passed.

I did not add any tests as I'm not very familiar with go tests and there were no existing tests for the `push.go` file. I don't feel as I have the skills needed to add a new test file and avoid running the `docker push` command for the test.

Here is a sample .yml file I used while testing:
```yaml
provider:
  name: faas
  gateway: http://localhost:8080

functions:
  movies:
    lang: node
    handler: ./movies
    image: burtonr/movies:0.1

  stronghash:
    skip_build: true
    image: alpine:latest
    fprocess: "sha512sum"

  ecs-cli:
    lang: Dockerfile
    handler: ./ecs-cli
    image: ecs-cli:latest
```

and the output (notice the "stronghash" function has `skip_build: true` and `ecs-cli` has no repository)
```bash
[0] % ./faas-cli push -f image-test.yml --parallel 3

Unable to push one or more of your functions to Docker Hub
 - ecs-cli

You must provide a username or registry prefix to the Function's image such as user1/function1
```

alexellis commented:
Hi @BurtonR - for this to be effective I think we need to block the entire flow if one or more of the functions fails the validation. In this example if we had half a dozen images the user may miss the error message which appears mid-flow - especially if in a parallel build.

Thoughts?

Alex

BurtonR commented:
I had considered that, but what happens when you have a function like the one I have in the example where it's just a base image? Would you have to comment that out in order for the push to happen?

I could update it to have the text in red... That would at least stand out and make it more obvious that something not-so-good happened. 

Or, another option could be to keep the error messages and display them all at the end, either just before, or just after the `worker done` message.

alexellis commented:
@BurtonR do you mean `stronghash`? The work item says that for `skip_build` you should bypass the validation.

This should be a completely blocking pre-validation.

```
$ faas-cli push
We were unable to push one or more of your functions to the Docker Hub.
- movie-lister
- ec2-autoscaler
- faas-cli

You must provide a username or registry prefix to the Function's image such as user1/function1.
If you need a Docker Hub account, you can sign up here: https://hub.docker.com
```

BurtonR commented:
Something like this?

Blocks the push completely. Also, note the "Skipping stronghash" on checking the `skip_build` property

![screenshot from 2018-02-19 11-45-28](https://user-images.githubusercontent.com/12687192/36391030-6b6a7ae6-156a-11e8-958e-194ac30c4cc4.png)

rgee0 commented:
@BurtonR lets also take the opportunity in the `unable` message to explain that the user will need to be logged in to docker hub.  

It may help head off the subsequent question, once the user has added their repo/prefix to the yaml

BurtonR commented:
@rgee0 

How does this look?

```
Unable to push one or more of your functions to Docker Hub
 - movies
 - ecs-cli

You must provide a username or registry prefix to the Function's image such as user1/function1.
Are you logged in to Docker Hub via "docker login"?
If you need a Docker Hub account, you can sign up here: https://hub.docker.com
```

rgee0 commented:
Keep in mind that it won’t always be a public push.  Are we picking up where the user is targeting with the push and tailoring?

I’d be tempted to merge the two Docker hub messages.  Something like:

`Also, ensure you are logged in to Docker hub - need an account? Create one here...`

alexellis commented:
I think our messages are quite succinct now so I think it makes sense to keep them consistent. How about we leave the login info until later if we really need it. Less is more in that way.

BurtonR commented:
So, what _should_ the message say? 

To be succinct, perhaps we should just leave it at the first line:
`You must provide a username or registry prefix to the function's image such as user1/function1`

The "sign up here" bit does feel pointed at beginners. Perhaps that would be better off in the docs only?

alexellis commented:
OpenFaaS has a core value of *developer-first* - and many of our developers may be using Docker for the first time by using OpenFaaS. 

If you feel leaving off `If you need a Docker Hub account, you can sign up here: https://hub.docker.com` is the right approach then let's do it. If we need to add it back at a later date then we can 👍

alexellis commented on commands/push.go:
```diff
@@ -95,3 +104,20 @@ func pushStack(services *stack.Services, queueDepth int) {
+		if function.SkipBuild {
+			continue
+		}
+		ip := "(\\b\\d{1,3}\\.\\d{1,3}\\.\\d{1,3}\\.\\d{1,3}\b)(:\\d+)?"
```
My gut feeling is that this is too complicated. Why are we using a `regex` when a string/contains would probably suffice?

```
if there's a `/` in the string.. then go for it?
```

BurtonR replied:
That sounds like a logical check... I was obviously over thinking this!

BurtonR commented:
I've updated the PR description as I've updated the check to look for `/` and updated the error message to a more succinct version (without the login or "sign up" text)

alexellis commented:
Thank you for updating the code @BurtonR - here's my next question.. what if someone applies the `--filter` flag in your example?

```
./faas-cli push -f image-test.yml --filter=movies
```

This should pass / push. Please could you confirm that it does?

alexellis commented on commands/push.go:
```diff
@@ -95,3 +107,17 @@ func pushStack(services *stack.Services, queueDepth int) {
+			continue
+		}
+
+		if !strings.Contains(function.Image, `/`) {
```
how about `&& !function.SkipBuild` ?

alexellis commented:
We're looking good now - the code is concise too. Would you be able to take a look at the testing done by @ericstoekl and @itscaro to see if you can cover the positive and negative cases with tests? If you would like some help feel free to ping in #contributors on Slack.

BurtonR commented:
@alexellis I've verified that the `--filter` option does function as expected.

On a `.yml` file in the PR description, where `movies` function _does_ have a repository, but the others do not, filtering by `--filter=movies` pushes the `movies` function successfully (does not check the other functions images).

alexellis commented:
💯

johnmccabe commented on commands/push.go:
```diff
@@ -95,3 +107,14 @@ func pushStack(services *stack.Services, queueDepth int) {
+	errMsg := make([]string, 0)
+	for name, function := range services.Functions {
+
+		if !function.SkipBuild && !strings.Contains(function.Image, `/`) {
```
Do you need to handle the case where the image only contains the `/`, for example `/imgfn`?

alexellis replied:
I don't think so? In that instance Docker will still error.

alexellis replied:
Thoughts?

BurtonR replied:
This is the error message that docker provides

```
invalid reference format
2018/02/22 09:21:09 ERROR - Could not execute command: [docker push /movies:0.1]
```

I think that's pretty clear, but do you think the `faas-cli` should provide something?

johnmccabe replied:
I just don't like the `invalid reference format` that results from the docker stderr when we are in a position to absorb it.

Its by no means a blocker however, just a personal preference for actionable output.

alexellis commented:
Please see the branch test_push_validation for an initial approach on unit testing the behaviour.

Collinshbk commented:
I am having the same issue. But all these back and forth but no clear solution or examples

alexellis commented:
The clear solution is that you need to log in.

@Collinshbk every eBook, training course and piece of official documentation explains that.

What solution are you wanting?
